File: src/backend/decoders/bgp.py
def decodeBGP(rawPayload):
    """
    Decode BGP messages from raw payload bytes.
    """
    import struct

    BGP_TYPES = {
        1: "OPEN",
        2: "UPDATE",
        3: "NOTIFICATION",
        4: "KEEPALIVE",
        5: "ROUTE-REFRESH",
    }
    BGP_ERRORS = {
        1: "Message Header Error",
        2: "OPEN Message Error",
        3: "UPDATE Message Error",
        4: "Hold Timer Expired",
        5: "Finite State Machine Error",
        6: "Cease",
    }
    try:
        if len(rawPayload) < 19:
            return None
        if rawPayload[:16] != b"\xff" * 16:
            return None
        msgLen = struct.unpack_from(">H", rawPayload, 16)[0]
        msgType = rawPayload[18]
        typeName = BGP_TYPES.get(msgType, f"Unknown({msgType})")
        result = {
            "Message Type": typeName,
            "bgp.type": typeName,
            "Message Length": msgLen,
            "bgp.length": msgLen,
        }
        if msgType == 1 and len(rawPayload) >= 29:
            version = rawPayload[19]
            asn = struct.unpack_from(">H", rawPayload, 20)[0]
            holdTime = struct.unpack_from(">H", rawPayload, 22)[0]
            routerId = ".".join(str(b) for b in rawPayload[24:28])
            result["BGP Version"] = version
            result["bgp.version"] = version
            result["ASN"] = asn
            result["bgp.asn"] = asn
            result["Hold Time"] = holdTime
            result["bgp.hold_time"] = holdTime
            result["Router ID"] = routerId
            result["bgp.router_id"] = routerId
        if msgType == 3 and len(rawPayload) >= 21:
            errCode = rawPayload[19]
            errSubcode = rawPayload[20]
            errName = BGP_ERRORS.get(errCode, f"Error {errCode}")
            result["Error Code"] = errCode
            result["bgp.error_code"] = errCode
            result["Error Name"] = errName
            result["bgp.error_name"] = errName
            result["Error Subcode"] = errSubcode
            result["bgp.error_subcode"] = errSubcode
        return result
    except Exception:
        return None

File: src/decoders/bgp.js
// Renders BGP packet details into the shared sidebar table UI.

const { createTable, dotField } = require("./shared");

function renderBgpTable(transportData) {
  const bgpData = transportData['BGP'];
  if (!bgpData) return;
  const bgpRows = [
    { name: 'Message Type', value: dotField(bgpData, 'bgp.type', 'Message Type') },
    { name: 'Message Length', value: dotField(bgpData, 'bgp.length', 'Message Length') },
  ];
  if (bgpData['BGP Version'] !== undefined) {
    bgpRows.push(
      { name: 'BGP Version', value: bgpData['BGP Version'] },
      { name: 'ASN', value: bgpData['ASN'] ?? '—' },
      { name: 'Hold Time', value: bgpData['Hold Time'] ?? '—' },
      { name: 'Router ID', value: bgpData['Router ID'] || '—' },
    );
  }
  if (bgpData['bgp.error_code'] !== undefined || bgpData['Error Code'] !== undefined) {
    bgpRows.push(
      { name: 'Error Name', value: dotField(bgpData, 'bgp.error_name', 'Error Name') },
      { name: 'Error Code', value: dotField(bgpData, 'bgp.error_code', 'Error Code') },
      { name: 'Error Subcode', value: dotField(bgpData, 'bgp.error_subcode', 'Error Subcode') },
    );
  }
  createTable(bgpRows, ['BGP Field', 'Value'], 'sidedatatable');
}

module.exports = { renderBgpTable };
